tsdate/date.py

```diff
@@ -993,41 +1051,33 @@ def backward_algorithm(
 
             edges = list(edges)
             for edge in edges:
-                dt = (backwards[edge.parent] *
-                      np.exp(np.subtract(forwards[edge.parent],
-                                         g_i[edge.child])))[matrix_indices.backward_dt]
+                dt = lls.expand_backwards(
+                    backwards[edge.parent] *
+                    np.exp(np.subtract(log_forwards[edge.parent], log_g_i[edge.child])))
                 span = edge.right - edge.left
                 if theta is not None and rho is not None:
                     b_l = (edge.left != 0)
                     b_r = (edge.right != ts.get_sequence_length())
                     ll_rec = np.power(
                         dt, b_l + b_r) * np.exp(-(dt * rho * span * 2))
-                    ll_mut = lls[mut_edges[edge.id], span][np.concatenate(
-                        matrix_indices.row_based_cols)]
-                    vv = dt * ll_mut * ll_rec
-                    vv = np.add.reduceat(vv, matrix_indices.col_based_cols)
+                    ll_mut = lls.get_mut_lik_colwise(edge)
+                    vv = lls.sum_tri_cols(dt * ll_mut * ll_rec)
                 elif theta is not None:
-                    ll_mut = lls[mut_edges[edge.id], span][np.concatenate(
-                        matrix_indices.row_based_cols)]
-                    vv = dt * ll_mut
-                    vv = np.add.reduceat(vv, matrix_indices.col_based_cols)
+                    ll_mut = lls.get_mut_lik_colwise(edge)
+                    vv = lls.sum_tri_cols(dt * ll_mut)
                 elif rho is not None:
                     b_l = (edge.left != 0)
                     b_r = (edge.right != ts.get_sequence_length())
                     ll_rec = np.power(
                         dt, b_l + b_r) * np.exp(-(dt * rho * span * 2))
-                    vv = dt * ll_rec
-                    vv = np.add.reduceat(vv, matrix_indices.col_based_cols)
+                    vv = lls.sum_tri_cols(dt * ll_rec)
                 else:
                     # Topology-only clock
-                    vv = np.add.reduceat(dt, matrix_indices.col_based_cols)
-            backwards[edge.child, 1:] = vv[1:]
-            norm[edge.child] = max(backwards[edge.child, :])
-            backwards[edge.child, :] = \
-                np.divide(backwards[edge.child, :], norm[edge.child])
-            backwards[edge.child, :][np.isnan(backwards[edge.child, :])] = 0
-    backwards_log = np.log(backwards)
-    posterior = np.exp(forwards + backwards_log)
+                    vv = lls.sum_tri_cols(dt)
+            backwards[edge.child, 1:] = vv[1:] / max(vv[1:])
+            backwards[edge.child, np.isnan(backwards[edge.child, :])] = 0
+    log_backwards = np.log(backwards)
```

Comment:
You can add  
`with np.errstate(divide='ignore'):`
above this line and it won't complain.

Reply:
Well, it's not so much that I am worried about h error, it's that I'm worried that I'm triggering it more often than before (I didn't have much of a feel for how much that was, though). I think it would be better to solve the error rather than hide it, though. You have a comment above that implies as much. What's the problem here?